### DISCUSSION TITLE
Is Hasura v3 / DDN OSS?

### DISCUSSION THREAD

**Discussion by mirzap** [hasura/graphql-engine#10556] [Question]
New documentation is confusing and is 100% SaaS focused. I can't figure out is Hasura (v3/DDN) still free to self host? If it is, is there any documentation how to do it?

**Comment by rakeshkky** [comment]
Hi @mirzap, the Hasura DDN API engine is already open-sourced and available in the graphql-engine repo [here](https://github.com/hasura/graphql-engine/tree/master/v3).

**Reply by sfcgeorge** [reply]
So graphql-engine V3 and the data connectors are open source, but as I understand it "DDN" is the cloud console GUI "thing" for creating metadata. Is that bit open source? It used to be part of graphql-engine but now looks like it's SaaS only?

Would graphql-engine V3 still run without connecting to cloud DDN? How would one create and import the metadata.json? There used to be metadata API but I can't find it in the docs.

**Reply by rakeshkky** [reply]
The DDN cloud console UI is a part of Hasura cloud offering and hence, it is not open sourced. AFAIK, one cannot author metadata in cloud console.

You've to use the [ddn](https://hasura.io/docs/3.0/cli/overview/) CLI for authoring metadata. Initialize a project as per instructions provided [here](https://hasura.io/docs/3.0/getting-started/quickstart/) and add your resources (data source, models etc.). To create a local build and serve engine locally, please follow the [local dev](https://hasura.io/docs/3.0/getting-started/build/build-your-api#step-1-create-a-supergraph-build) workflow.

Tip: To use locally built v3-engine OSS.
- build a docker image using the [Dockerfile](https://github.com/hasura/graphql-engine/blob/master/v3/Dockerfile)
- overwrite the image name under `dockerfile_inline` in your DDN project's `compose.yaml` file.
And run `ddn run docker-start` and `ddn console --local` commands.

**Reply by rakeshkky** [reply]
> There used to be metadata API but I can't find it in the docs.

Hasura v3/DDN doesn't have a metadata API.

**Reply by mirzap** [reply]
I tried running docker-compose in the v3 folder. I get two containers running, but there is no Hasura UI, and I have no idea how to start. I’ve been working with Hasura since v1, and this is quite disappointing.

If I get this right, is there no more UI interface for building tables, APIs, etc? 

I’ve long avoided Supabase because I like Hasura; it is a great product. But now I’m considering switching completely. They improve and deliver constantly, while hasura has been stagnating for some time now.

**Reply by hgiasac** [reply]
> If I get this right, is there no more UI interface for building tables, APIs, etc?

@mirzap you are correct. The long-term goal of Hasura v3 is to support many databases. Each database has a different API spec so we no longer take much effort for database migrations. If you only need to connect Hasura with PostgreSQL Hasura v2 is still very good to be used.  

I recommend you download the DDN CLI to create the boilerplate and follow the guide in the Hasura v3 docs. The DX of OSS and cloud are the same. If you don't want to deploy the project to the Hasura cloud you can build the metadata locally and run the stack with docker-compose or Kubernetes. 

https://hasura.io/docs/3.0/getting-started/quickstart (You can stop at step 7 and browse http://localhost:3000 to run the Graphiql playground)

For metadata development, we maintain the CLI, Language Server Protocol, and VSCode IDE so developers can develop metadata directly in the editor. The downside of web UI development in v2 is that the performance is slow if the metadata is big. The IDE will be more lightweight to run.

Or you can use Hasura v2 as a migration tool to create database migrations and develop metadata in v3. It's a bit tricky but still works well.

**Reply by fungiboletus** [reply]
I couldn't find the source code of the DDN CLI. Is it open-source or proprietary?

I tried to find the source code of the Language Server Protocol and it seems to be located in this repository: https://github.com/hasura/hasura-lsp-server according to the VSCode plugin metadata. But this is not public, is it open-source or proprietary?

I'm realising that Hasura v3 may require the use of non open-source components. I understand that living well from open-source software is challenging, I don't have a good solution. But I personally have a policy that consists of not relying on proprietary software.

So long, and thanks for all the fish.

**Reply by hgiasac** [reply]
@fungiboletus Currently, the DDN CLI, Web console, and LSP are proprietary. The following components are open source with Apache license:
- [Engine v3](https://github.com/hasura/graphql-engine/tree/master/v3).
- Data Connectors.

Most DDN CLI, Web console, and LSP features are for dev tools and the DDN cloud. They won't lock you to run self-hosted environments in production. Therefore you can use Hasura v3 without worrying about the license.

**Reply by fungiboletus** [reply]
Thanks for the clarification. I understand that I can self host the engine, but I don't like to rely on development tools that cannot be distributed, tweaked, fixed, or replaced.

I also think that having those necessary tools under a proprietary license makes the open-source part of the platform a lot less attractive to open-source contributions, but I may be wrong. I also don't know how much open-source contributions you got anyway.

**Reply by azzeddineMadi-cirium** [reply]
> > If I get this right, is there no more UI interface for building tables, APIs, etc?
> 
> @mirzap you are correct. The long-term goal of Hasura v3 is to support many databases. Each database has a different API spec so we no longer take much effort for database migrations. If you only need to connect Hasura with PostgreSQL Hasura v2 is still very good to be used.
> 
> I recommend you download the DDN CLI to create the boilerplate and follow the guide in the Hasura v3 docs. The DX of OSS and cloud are the same. If you don't want to deploy the project to the Hasura cloud you can build the metadata locally and run the stack with docker-compose or Kubernetes.
> 
> https://hasura.io/docs/3.0/getting-started/quickstart (You can stop at step 7 and browse http://localhost:3000 to run the Graphiql playground)
> 
> For metadata development, we maintain the CLI, Language Server Protocol, and VSCode IDE so developers can develop metadata directly in the editor. The downside of web UI development in v2 is that the performance is slow if the metadata is big. The IDE will be more lightweight to run.
> 
> Or you can use Hasura v2 as a migration tool to create database migrations and develop metadata in v3. It's a bit tricky but still works well.

Regarding following the QuickStart, Step 1 is connecting to Hasura cloud, which I don't want to because I want to sandbox this work, how to avoid that, and continue with the initialising of the Supergraph?
Thanks

**Reply by mirzap** [reply]
> > > If I get this right, is there no more UI interface for building tables, APIs, etc?
> > 
> > 
> > @mirzap you are correct. The long-term goal of Hasura v3 is to support many databases. Each database has a different API spec so we no longer take much effort for database migrations. If you only need to connect Hasura with PostgreSQL Hasura v2 is still very good to be used.
> > I recommend you download the DDN CLI to create the boilerplate and follow the guide in the Hasura v3 docs. The DX of OSS and cloud are the same. If you don't want to deploy the project to the Hasura cloud you can build the metadata locally and run the stack with docker-compose or Kubernetes.
> > https://hasura.io/docs/3.0/getting-started/quickstart (You can stop at step 7 and browse http://localhost:3000 to run the Graphiql playground)
> > For metadata development, we maintain the CLI, Language Server Protocol, and VSCode IDE so developers can develop metadata directly in the editor. The downside of web UI development in v2 is that the performance is slow if the metadata is big. The IDE will be more lightweight to run.
> > Or you can use Hasura v2 as a migration tool to create database migrations and develop metadata in v3. It's a bit tricky but still works well.
> 
> Regarding following the QuickStart, Step 1 is connecting to Hasura cloud, which I don't want to because I want to sandbox this work, how to avoid that, and continue with the initialising of the Supergraph? Thanks

You can't. Hasura v3 is NOT open source, no matter how they try to sugarcoat it with open sourcing few components. Key component - build/metadata engine is closed-source, cloud-only. You can't build metadata without using cloud. Which makes all other "open source" components useless.Too bad, I gave up of Hasura. I have one more large project using v2, but planning a large migration to Supabase in a  year or two.

**Reply by azzeddineMadi-cirium** [reply]
> > > > If I get this right, is there no more UI interface for building tables, APIs, etc?
> > > 
> > > 
> > > @mirzap you are correct. The long-term goal of Hasura v3 is to support many databases. Each database has a different API spec so we no longer take much effort for database migrations. If you only need to connect Hasura with PostgreSQL Hasura v2 is still very good to be used.
> > > I recommend you download the DDN CLI to create the boilerplate and follow the guide in the Hasura v3 docs. The DX of OSS and cloud are the same. If you don't want to deploy the project to the Hasura cloud you can build the metadata locally and run the stack with docker-compose or Kubernetes.
> > > https://hasura.io/docs/3.0/getting-started/quickstart (You can stop at step 7 and browse http://localhost:3000 to run the Graphiql playground)
> > > For metadata development, we maintain the CLI, Language Server Protocol, and VSCode IDE so developers can develop metadata directly in the editor. The downside of web UI development in v2 is that the performance is slow if the metadata is big. The IDE will be more lightweight to run.
> > > Or you can use Hasura v2 as a migration tool to create database migrations and develop metadata in v3. It's a bit tricky but still works well.
> > 
> > 
> > Regarding following the QuickStart, Step 1 is connecting to Hasura cloud, which I don't want to because I want to sandbox this work, how to avoid that, and continue with the initialising of the Supergraph? Thanks
> 
> You can't. Hasura v3 is NOT open source, no matter how they try to sugarcoat it with open sourcing few components. Key component - build/metadata engine is closed-source, cloud-only. You can't build metadata without using cloud. Which makes all other "open source" components useless.Too bad, I gave up of Hasura. I have one more large project using v2, but planning a large migration to Supabase in a year or two.

Can anyone from Hasura confirm that please, I really don't want to invest more time on it if they are forcing people to connect to their cloud.
